src/bots/marketBidAskTwapCrank.ts

```diff
@@ -0,0 +1,244 @@
+import {
+	DLOB,
+	DriftClient,
+	DriftEnv,
+	UserMap,
+	SlotSubscriber,
+	MarketType,
+	PositionDirection,
+	getUserStatsAccountPublicKey,
+	promiseTimeout,
+} from '@drift-labs/sdk';
+import { Mutex } from 'async-mutex';
+
+import { logger } from '../logger';
+import { Bot } from '../types';
+import { BaseBotConfig } from '../config';
+import {
+	TransactionSignature,
+	VersionedTransaction,
+	AddressLookupTableAccount,
+} from '@solana/web3.js';
+import { ConfirmOptions, Signer } from '@solana/web3.js';
+
+export async function sendVersionedTransaction(
+	driftClient: DriftClient,
+	tx: VersionedTransaction,
+	additionalSigners?: Array<Signer>,
+	opts?: ConfirmOptions,
+	timeoutMs = 5000
+): Promise<TransactionSignature | null> {
+	// @ts-ignore
+	tx.sign((additionalSigners ?? []).concat(driftClient.provider.wallet.payer));
+
+	if (opts === undefined) {
+		opts = driftClient.provider.opts;
+	}
+
+	const rawTransaction = tx.serialize();
+	let txid: TransactionSignature;
+	try {
+		txid = await promiseTimeout(
+			driftClient.provider.connection.sendRawTransaction(rawTransaction, opts),
+			timeoutMs
+		);
+	} catch (e) {
+		console.error(e);
+		throw e;
+	}
+
+	return txid;
+}
+
+export class MarketBidAskTwapCrank implements Bot {
+	public readonly name: string;
+	public readonly dryRun: boolean;
+	public readonly runOnce: boolean;
+	public readonly defaultIntervalMs: number = 600000;
+
+	private slotSubscriber: SlotSubscriber;
+	private driftClient: DriftClient;
+	private intervalIds: Array<NodeJS.Timer> = [];
+	private userMap: UserMap;
+	private driftEnv: DriftEnv;
+
+	private dlob: DLOB;
+	private latestDlobSlot: number;
+	private lastDlobRefreshTime = 0;
+	private lookupTableAccount: AddressLookupTableAccount;
+	private pollingIntervalMs: number;
+
+	private watchdogTimerMutex = new Mutex();
+	private watchdogTimerLastPatTime = Date.now();
+
+	constructor(
+		driftClient: DriftClient,
+		slotSubscriber: SlotSubscriber,
+		driftEnv: DriftEnv,
+		config: BaseBotConfig
+	) {
+		this.slotSubscriber = slotSubscriber;
+		this.name = config.botId;
+		this.dryRun = config.dryRun;
+		this.runOnce = config.runOnce || false;
+		this.driftClient = driftClient;
+		this.driftEnv = driftEnv;
+	}
+
+	public async init() {
+		logger.info(`${this.name} initing`);
+		this.lookupTableAccount =
+			await this.driftClient.fetchMarketLookupTableAccount();
+
+		// initialize userMap instance
+		this.userMap = new UserMap(
```

Comment:
are the user maps every shared btwn bots?

Reply:
not sure..